components/raftstore/src/store/worker/pd.rs

```diff
@@ -1290,6 +1334,43 @@ where
         };
         self.remote.spawn(f);
     }
+
+    fn handle_region_cpu_records(&mut self, records: Arc<CpuRecords>) {
+        calculate_region_cpu_records(self.store_id, records, &mut self.region_cpu_records)
+    }
+}
+
+fn calculate_region_cpu_records(
+    store_id: u64,
+    records: Arc<CpuRecords>,
+    region_cpu_records: &mut HashMap<u64, (Vec<u64>, Vec<u32>, u32)>,
+) {
+    let timestamp_secs = records.begin_unix_time_secs;
+
+    for (tag, ms) in &records.records {
+        let record_store_id = tag.infos.store_id;
+        let region_id = &tag.infos.region_id;
+        let extra_attachment = &tag.infos.extra_attachment;
+        if extra_attachment.is_empty() || record_store_id != store_id {
+            continue;
+        }
+
+        let ms = *ms as u32;
+        match region_cpu_records.get_mut(region_id) {
+            Some((ts, cpu_time, total)) => {
+                if *ts.last().unwrap() == timestamp_secs {
+                    *cpu_time.last_mut().unwrap() += ms;
+                } else {
+                    ts.push(timestamp_secs);
```

Comment:
Does this will have too many entries and cost high memory?

Reply:
It will be cleaned up every time we report it through the region heartbeat. Maybe GC logic is needed also? WDYT?

Reply:
> Does this will have too many entries and cost high memory?

I think we can maintain a top-k map here rather than saving them all.

Reply:
Seems every second will have an entry. We only care about the tags of `region_id`, so we can sum them up in the heartbeat interval instead of storing them all.

Reply:
Now it will only store the total CPU time since the last region heartbeat. PTAL.